components/screens/ObservationsScreen.tsx

```diff
@@ -46,6 +60,42 @@ const ObservationsListScreen = ({route}: NativeStackScreenProps<ObservationsStac
   );
 };
 
+/**
+ * Return the previous screen from the route stack.
+ */
+const lastScreen = (state: StackNavigationState<ObservationsStackParamList>) => state.routes[state.index - 1];
+
+const ObservationImageEditScreen = ({route, navigation}: ObservationStackProps<'observationEditImage'>) => {
+  const {image, caption} = route.params;
+
+  const returnToScreen = lastScreen(navigation.getState());
+
+  const handleSetCaption = useCallback(
+    (caption: string) => {
+      switch (returnToScreen.name) {
+        case 'observationSubmit': {
+          // Ideally the entire route would refine types based os the returnToScreen.name refinement
+          // however that doesn't seem to be the case when using the getState() records.
+          //
+          // We are trusting that the routes are all contain the correct params and refining based
+          // on the name
+          const submitRoute = returnToScreen as ObservationStackProps<'observationSubmit'>['route'];
+          navigation.navigate(submitRoute.name, {
+            ...submitRoute.params,
+            set_image_caption: [image, caption],
+          });
+          break;
+        }
+        default: {
+          // noop
```

Comment:
The implication of this is that the app gets stuck in this view without a way to navigate back, right?

Reply:
Yeah, my thinking here for the `default` case is to either `throw` or use `.goBack()`. It could also `throw` in development builds and `goBack` in production ones.

Reply:
Another option is to use a function as a param. This works but `react-navigation` warns that it's not serializable so doesn't work with deep linking, which I don't think we'd ever use for this screen anyway.

Reply:
More docs on how `react-navigation` thinks this should be handled https://reactnavigation.org/docs/params#passing-params-to-a-previous-screen

You _can_ use a callback function and it works. This is the warning you're given

```
 WARN  Non-serializable values were found in the navigation state. Check:

Observations > observationEditImage > params.onSetImageCaption (Function)

This can break usage such as persisting and restoring state. This might happen if you passed non-serializable values such as function, class instances etc. in params. If you need to use components with callbacks in your options, you can use 'navigation.setOptions' instead. See https://reactnavigation.org/docs/troubleshooting#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state for more details. 
```